File: backend/app/models/recipe.rb
# frozen_string_literal: true

##
# == Fields
# name:             string
# category:         ApplicationRecord::Category
# user:             ApplicationRecord::User
# preparation_time: integer
# cooking_time:     integer
# number_of_people: integer
# difficulty:       string
# price:            string
# image:            string
# preparations:     ApplicationRecord::RecipesPreparation
# description:      text
# created_at:       datetime
# updated_at:       datetime
class Recipe < ApplicationRecord

  belongs_to :user
  belongs_to :category
  has_many :ingredients_recipes, dependent: :destroy
  has_many :ingredients, through: :ingredients_recipes
  has_many :recipes_preparations, dependent: :destroy
  has_and_belongs_to_many :kitchenwares
  has_and_belongs_to_many :countries
  has_many :recipes_images, dependent: :destroy

  accepts_nested_attributes_for :recipes_preparations, :ingredients_recipes, allow_destroy: true

  validates :name, presence: true, length: { maximum: 100 }
  validates :preparation_time,
            :cooking_time,
            :number_of_people,
            allow_nil: true, numericality: { greater_than: 0 }
  enum difficulty: %i[easy normal hard], _prefix: :difficulty
  enum price: %i[low medium high], _suffix: true
  validates_presence_of :image

end

File: frontend/src/entity/recipe.ts
import type Category from './category'
import type Country from './country'
import type { IngredientRecipe } from './ingredient'
import type Kitchenware from './kitchenware'
import type User from './user'

export default interface Recipe {
  id: number
  image: string
  name: string
  countryId: number | null
  categoryId: number
  userId: number
  preparationTime: number | null
  cookingTime: number | null
  numberOfPeople: number | null
  difficulty: RecipeDifficulty
  price: RecipePrice
  createdAt?: Date
  updatedAt?: Date
  description: string

  category?: Partial<Category>
  user?: Partial<User>
  country?: Partial<Country>

  kitchenwares?: Kitchenware[]
  ingredients?: IngredientRecipe[]
  instructions?: InstructionsRecipe[]
  // Create or update a recipe
  kitchenwareIds?: Array<number>
  ingredientsRecipesAttributes?: Array<IngredientsRecipesAttributes>
  instructionsRecipesAttributes?: Array<InstructionsRecipesAttributes>
}

export interface InstructionsRecipe {
  id: number
  step: number
  comment: string
}

export type RecipeDifficulty = 'easy' | 'normal' | 'hard'
export type RecipePrice = 'low' | 'medium' | 'high'

export interface IngredientsRecipesAttributes {
  id?: number
  ingredientId: number
  quantity: number | null
  measureId: number | null
  comment: string | null
  _destroy?: '1' | true
}

export interface InstructionsRecipesAttributes {
  id?: number
  step: number
  comment: string
  _destroy?: '1' | true
}
